components/autofill/src/api/credit_cards.rs

```diff
@@ -0,0 +1,594 @@
+    let tx = conn.transaction()?;
+    let mut credit_cards = Vec::new();
+    let sql = format!(
+        "SELECT
+            {common_cols}
+        FROM credit_cards_data",
+        common_cols = CREDIT_CARD_COMMON_COLS
+    );
+
+    {
+        let mut stmt = tx.prepare(&sql)?;
+        let credit_cards_iter = stmt.query_map(NO_PARAMS, |row| Ok(CreditCard::from_row(row)?))?;
+
+        for credit_cards_result in credit_cards_iter {
+            credit_cards.push(credit_cards_result.expect("Should unwrap credit_card"));
+        }
+    }
+
+    tx.commit()?;
+    Ok(credit_cards)
+}
+
+#[allow(dead_code)]
+pub fn update_credit_card(conn: &mut Connection, credit_card: CreditCard) -> Result<()> {
+    let tx = conn.transaction()?;
+    tx.execute_named(
+        "UPDATE credit_cards_data
+        SET cc_name                     = :cc_name,
+            cc_number                   = :cc_number,
+            cc_exp_month                = :cc_exp_month,
+            cc_exp_year                 = :cc_exp_year,
+            cc_type                     = :cc_type,
+            time_created                = :time_created,
+            time_last_used              = :time_last_used,
+            time_last_modified          = :time_last_modified,
+            times_used                  = :times_used,
+            sync_change_counter         = sync_change_counter + 1
+        WHERE guid                      = :guid",
+        rusqlite::named_params! {
+            ":cc_name": credit_card.fields.cc_name,
+            ":cc_number": credit_card.fields.cc_number,
+            ":cc_exp_month": credit_card.fields.cc_exp_month,
+            ":cc_exp_year": credit_card.fields.cc_exp_year,
+            ":cc_type": credit_card.fields.cc_type,
+            ":time_created": credit_card.time_created,
+            ":time_last_used": credit_card.time_last_used,
+            ":time_last_modified": credit_card.time_last_modified,
+            ":times_used": credit_card.times_used,
+            ":guid": credit_card.guid,
+        },
+    )?;
+
+    tx.commit()?;
+    Ok(())
+}
+
+#[allow(dead_code)]
+pub fn delete_credit_card(conn: &mut Connection, guid: &Guid) -> Result<bool> {
+    let tx = conn.transaction()?;
+
+    // check that guid exists
+    let exists = tx.query_row(
+        "SELECT EXISTS (
+            SELECT 1
+            FROM credit_cards_data d
+            WHERE guid = :guid
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM   credit_cards_tombstones t
+                    WHERE  d.guid = t.guid
+                )
+        )",
+        &[guid.as_str()],
+        |row| row.get(0),
+    )?;
+
+    if exists {
+        tx.execute_named(
+            "DELETE FROM credit_cards_data
+            WHERE guid = :guid",
+            rusqlite::named_params! {
+                ":guid": guid.as_str(),
+            },
+        )?;
+
+        tx.execute_named(
```

Comment:
Once the Sync mirror is in place, I think we'll want to change this to only insert tombstones for credit cards that are in the mirror.

It doesn't hurt to insert tombstones for things that aren't in the mirror—except use a tiny bit more space on the server—it's just that those tombstones won't mean anything. If the card isn't in the mirror, we know it hasn't been synced yet, so it doesn't exist on the server or other devices.

Reply:
Good to know! I'll go ahead and update this logic know ahead of the sync work.